contracts/BonusRewards.sol

```diff
@@ -3,11 +3,340 @@
  * @author crypto-pumpkin
+ * @notice ETH is not allowed to be an bonus token, use wETH instead
  */
-contract BonusRewards is Ownable {
+contract BonusRewards is IBonusRewards, Ownable, ReentrancyGuard {
+  using SafeERC20 for IERC20;
+
+  uint256 private constant WEEK = 7 days;
+  uint256 private constant CAL_MULTIPLIER = 1e12; // help calculate rewards/bonus PerToken only. 1e12 will allow meaningful $1 deposit in a $1bn pool
+  bool public paused;
+  address[] private responders;
+  address[] private poolList;
+  // lpToken => Pool
+  mapping(address => Pool) private pools;
+  // lpToken => User address => User data
+  mapping(address => mapping(address => User)) private users;
+  // lpToken => bonus token => [] allowed authorizers to add bonus tokens
+  mapping(address => mapping(address => address[])) private allowedTokenAuthorizers;
+  // bonusTokenAddr => 1, used to avoid collecting bonus token when not ready
+  mapping(address => uint8) private bonusTokenAddrMap;
+
+  modifier notPaused() {
+    require(!paused, "BonusRewards: paused");
+    _;
+  }
+
+  function getPoolList() external view override returns (address[] memory) {
+    return poolList;
+  }
+
+  function getPool(address _lpToken) external view override returns (Pool memory) {
+    return pools[_lpToken];
+  }
+
+  function viewRewards(address _lpToken, address _user) public view override returns (uint256[] memory) {
+    Pool memory pool = pools[_lpToken];
+    User memory user = users[_lpToken][_user];
+    uint256[] memory rewards = new uint256[](pool.bonuses.length);
+    if (user.amount <= 0) return rewards;
+
+    uint256 rewardsWriteoffsLen = user.rewardsWriteoffs.length;
+    for (uint256 i = 0; i < rewards.length; i ++) {
+      Bonus memory bonus = pool.bonuses[i];
+      if (bonus.startTime < block.timestamp && bonus.remBonus > 0) {
+        uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+        uint256 bonusForTime = _calRewardsForTime(bonus, pool.lastUpdatedAt);
+        uint256 bonusPerToken = bonus.accRewardsPerToken + bonusForTime / lpTotal;
+        uint256 rewardsWriteoff = rewardsWriteoffsLen <= i ? 0 : user.rewardsWriteoffs[i];
+        rewards[i] = user.amount * bonusPerToken / CAL_MULTIPLIER - rewardsWriteoff;
+      }
+    }
+    return rewards;
+  }
+
+  function getUser(address _lpToken, address _account) external view override returns (User memory, uint256[] memory) {
+    return (users[_lpToken][_account], viewRewards(_lpToken, _account));
+  }
+
+  function getAuthorizers(address _lpToken, address _bonusTokenAddr) external view override returns (address[] memory) {
+    return allowedTokenAuthorizers[_lpToken][_bonusTokenAddr];
+  }
+
+  function getResponders() external view override returns (address[] memory) {
+    return responders;
+  }
+
+  /// @notice update pool's bonus per staked token till current block timestamp
+  function updatePool(address _lpToken) public override {
```

Comment:
I would consider making this nonreentrant for external use. I can't think of an exploit but given you are allowing third party tokens to be added I think it is wise that all external calls cannot be reentered

Reply:
Converting this to private since it doesn't need to be public anymore without the COVER rewards weights update.